keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,294 @@
+
+In addition to the improvements to scale and performance, we aim to resolve a
+specific problem. The long standing "stale read" issue
+(https://github.com/kubernetes/kubernetes/issues/59848) remains open because
+reflectors default to resourceVersion=”0” for their initial list requests. If
+the reflectors instead use a consistent read for their initial list request,
+they could not "going back in time" when components are restarted and this issue
+would be solved. "Going back in time" can curently happen if the initial list
+request is served from a stale watch cache with data much older than the
+reflector has previously observed or if the api-server or etcd are partitioned.
+
+We have held off on switching reflectors to using consistent read for the
+initial list, even though we know it is more correct, due to concerns with the
+impact on large scale use cases. But if we serve consistent reads from cache,
+there would be very little difference in scalability to how the kube-apiserver
+serves the resourceVersion="0" list requests from reflectors today.
+
+### Goals
+
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scalability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
+
+## Proposal
+
+### Leveraging the Progress Notify Mechanism
+
+Guard this by a `WatchCacheConsistentReads` feature gate.
+
+#### Alternative 1: Use WithProgressNotify to enable automatic watch updates
+
+Create etcd watches with `WithProgressNotify` enabled (available in all etcd 3.x versions).
+
+When `WithProgressNotify` is enabled on an etcd watch, etcd sends progress
+events to the watch automatically. By default etcd sends progress events every
+10 minutes, which is not frequent enough to be useful for our needs, so we will
+modify etcd to send them must more frequently, e.g. every 20ms.
+
+When an consistent read request is received and the watch cache is enabled:
+
+- Get the current revision from etcd for the resource type being served. The returned revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- Use the existing `waitUntilFreshAndBlock` function in the watch cache to wait briefly (20ms?) for the watch to catch up to the current revision.
+- If the block times out, skip the cache and serve the request directly from storage (etcd).
+
+To get the revsion we have two options: 
+
+- Use an etcd range request with `WithCount` enabled so etcd return only a count and revision 
+- Use an etcd range request against a known empty range with limit=1 as an additional guard (since etcd does not allow for limit=0)
+
+We will benchmark and select the most efficient of these options.
+
+#### Alternative 2: Use WatchProgressRequest to request watch updates when needed
+
+etcd 3.4+ provides a `WatchProgressRequest` request that can be made on a watch channel. When requested,
```

Comment:
What happens if we send a "WatchProgressRequest" to an etcd version that doesn't understand it?

Reply:
etcd responds with a `NOT_FOUND` (https://github.com/grpc/grpc/blob/master/doc/statuscodes.md) code.